theano/gpuarray/tests/check_dnn.py

```diff
@@ -0,0 +1,637 @@
+
+
+def dnn_gradweight(img, topgrad, kerns_shp, alpha=1, beta=0, out=None, border_mode='valid', subsample=(1, 1),
+                   dilation=(1, 1), conv_mode='conv', algo=None, precision=None):
+    ctx_name = infer_context_name(img, topgrad)
+
+    img = as_gpuarray_variable(img, ctx_name)
+    topgrad = as_gpuarray_variable(topgrad, ctx_name)
+
+    img = gpu_contiguous(img)
+    topgrad = gpu_contiguous(topgrad)
+
+    kerns_shp = theano.tensor.as_tensor_variable(kerns_shp)
+    precision = get_precision(precision, [img, topgrad])
+
+    desc = GpuDnnConvDesc(border_mode=border_mode, subsample=subsample, dilation=dilation,
+                          conv_mode=conv_mode, precision=precision)(kerns_shp)
+    if beta == 0:
+        real_out = GpuAllocEmpty(dtype=img.dtype, context_name=ctx_name)(*kerns_shp)
+    else:
+        assert out is not None
+        out = as_gpuarray_variable(out, ctx_name)
+        out = gpu_contiguous(out)
+        check = Assert('GpuDnnConvGradW: qiven output (for beta not null) does not have expected shape')
+        real_out = check(out, theano.tensor.all(theano.tensor.eq(out.shape, kerns_shp)))
+
+    return GpuDnnConvGradW(algo=algo)(img, topgrad, real_out, desc, alpha, beta)
+
+
+def dnn_gradinput(kerns, topgrad, img_shp, alpha=1, beta=0, out=None, border_mode='valid', subsample=(1, 1),
+                  dilation=(1, 1), conv_mode='conv', algo=None, precision=None):
+    ctx_name = infer_context_name(kerns, topgrad)
+
+    kerns = as_gpuarray_variable(kerns, ctx_name)
+    topgrad = as_gpuarray_variable(topgrad, ctx_name)
+
+    kerns = gpu_contiguous(kerns)
+    topgrad = gpu_contiguous(topgrad)
+
+    img_shp = theano.tensor.as_tensor_variable(img_shp)
+    precision = get_precision(precision, [kerns, topgrad])
+
+    desc = GpuDnnConvDesc(border_mode=border_mode, subsample=subsample, dilation=dilation,
+                          conv_mode=conv_mode, precision=precision)(kerns.shape)
+    if beta == 0:
+        real_out = GpuAllocEmpty(dtype=kerns.dtype, context_name=ctx_name)(*img_shp)
+    else:
+        assert out is not None
+        out = as_gpuarray_variable(out, ctx_name)
+        out = gpu_contiguous(out)
+        check = Assert('GpuDnnConvGradI: qiven output (for beta not null) does not have expected shape')
+        real_out = check(out, theano.tensor.all(theano.tensor.eq(out.shape, img_shp)))
+
+    return GpuDnnConvGradI(algo=algo)(kerns, topgrad, real_out, desc, alpha, beta)
+
+
+class BaseGpuDnnConvChooseAlgo(DnnBase):
+    """
+    This class and its subclasses allow to retrieve a cuDNN algorithm
+    at runtime without any computation, given the user choose option
+    (time_once, time_on_shape_change, guess_once or guess_on_shape_change).
+    To help reduce whole test time, I suggest we use these classes when
+    algo is one of choose options, as any chosen algorithm would have
+    been tested by the other exhaustive tests.
+    """
```

Comment:
The goal was to reduce whole execution time for this script.

As the script will test all algorithms one by one, then an algo chosen at runtime (when using `time_once`, `guess_once`, etc.) will be also tested by the script, in any case. So I considered that we don't need to explicitely run convolutions when we choose algo at runtime (otherwise, we will have 2 runs for a same algo with the same test case): we just want to check if algo is "well chosen".

But indeed, to check if an algo is "well chosen", we may need to run it. I thought we could also get it from other tests, but it would imply to force an order of execution for the tests.

As it is quite experimental, I can remove it.